keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,577 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Liveness probes currently use the `terminationGracePeriodSeconds` on both
+normal shutdown and when probes fail. Hence, if a long termination period is
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to probes, `probe.terminationGracePeriodSeconds`.
+When set, it will override the `terminationGracePeriodSeconds` for liveness or
+startup termination, and will be ignored for readiness probes. This maintains
+the current behaviour if desired while providing configuration to address this
+unintended behaviour.
+
+## Motivation
+
+This change is important to give users the option to decouple these fields, as
+the current behaviour is counter-intuitive and doesn't provide an option to
+decouple.
+
+[Other implementation options](#alternatives) might avoid an API change, but
+would end up breaking backwards compatibility which some users may have come to
+[expect/rely on](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756100727),
+even if we take a phased approach to introduce them.
+
+An example of a scenario where this bug was apparent is a long outage reported
+by an ingress controller:
+
+> the controller uses a 3600s grace period so that it can drain long requests
+> from end user app traffic (it listens on endpoints that can remain available
+> for hours like host network, but also incoming tcp load balancers which have
+> open connections), but when the controller wedges it takes an hour to
+> resolve, which is the worst possible outcome (the process wedges immediately,
+> but kube has to wait the full hour to kill it, defeating the goal of
+> liveness). [comment](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756201502)
+
+### Goals
+
+- Liveness probes can have a configurable timeout on failure.
+- Existing behaviour is not changed.
```

Comment:
is there any chance to get a race between the termination on node shutdown and liveness probe failure. Let's say liveness probe is set up to be extremely sensitive to the termination signal that pod receives. So pod started termination because it received sig kill, liveness probe immediately failed, and instead of regular interval, pod assumes liveness probe's interval override. So depending whether liveness probe checked in the period between termination started and finished, grace period will be different.

Reply:
The termination paths for pod deletion/node drain and liveness probe failures are sort of separate (they use the same underlying function calls to kill the containers, but pod deletion always includes a hard override for the termination grace period). I don't think this would introduce any greater chance of a race condition being encountered.

Is it possible right now for a liveness probe to trigger specific container termination during a pod already in process of terminating? That's the only scenario I can think of that might pose an issue.

Reply:
@matthyx is fixing this exact scenario, so it will no longer be possible: https://github.com/kubernetes/kubernetes/pull/98571

Reply:
agree we should not probe for startup/liveness when a pod is undergoing graceful termination.  fwiw, if you send multiple stop calls to a container today, both primary runtimes hold a lock that prevents the subsequent call from running or reducing the termination period.  this is being fixed as well https://github.com/kubernetes/kubernetes/pull/98507/files